### PR TITLE
Fix: RespondWorkflowTaskCompleted should the workflow termination event into the same event batch as the workflowtask failure event

### PR DESCRIPTION
**Pull request by lina-temporal**
## What changed?
We [previously updated RespondWorkflowTaskCompleted](https://github.com/temporalio/temporal/pull/6180) to terminate on certain non-retryable errors (payload size exceeded) instead of fail. Termination indicates that the failure was sourced from the service, rather than the customer code/SDK. The change inadvertedly resulted in the wrong `CompletionEventBatchID` being written into mutable state, since both `failWorkflowTask` and `TerminateWorkflow` allocate a new event batch. `CompletionEventBatchID` should be set to the batch started in `failWorkflowTask`, but it was overridden from `TerminateWorkflow`. 

When replication to another cluster was involved, or the events cache is otherwise evicted, `MutableState.GetCompletionEvent()` was no longer able to load the failed event, as the `CompletionEventBatchID` used to load from persistence is past the WFT failure event. This caused active transfer tasks involving the terminated workflow to be sent to the DLQ on secondary/standby clusters.

This fix:
- correctly writes the right event batch ID into the workflow termination event by using `AddWorkflowExecutionTerminatedEvent` over `TerminateWorkflow`
- updates `GetCompletionEvent()` to handle internal errors for terminated workflows by iterating through the last batch of events via persistence

## How did you test it?
Set up a local temporal server, and a simple test worker/workflow. The test workflow simply launches a child workflow and awaits the result; the child workflow returns an output with a large (2MB) payload, to trigger termination. 

✅ Validating test case:
- start with unpatched temporal
- run test workflow, observe termination
- run `tdbg w d` and observe an incorrect `completionEventBatchId`
- restart server
- run `tdbg w rt` to refresh tasks
- observe `"error":"unable to get workflow completion event"` message in logs

✅  Validating backfill for corrupted `completionEventBatchId` workflows:
- (continue from "Validated test case" steps)
- stop unpatched temporal server
- start patched temporal server
- run `tdbg w rt` to refresh tasks
- observe server no longer shows error message/makes progress

✅  Validating newly-terminated workflows:
- start with patched temporal server
- run test workflow, observe termination
- run `tdbg w d` and observe the correct `completionEventBatchId`, ensures it matches the `WorkflowTaskFailed` event ID

## Potential risks
- introduces new reverse-history search logic into `GetCompletionEvent()` for edge cases which previously wasn't there; tried to mitigate impact by only requesting a single page + batch of results
- once again changes the `CompletionEventBatchID` for workflow termination events, which may cause transfer task errors if incorrect

## Is hotfix candidate?
Yes

### CODE REVIEW

File: service/history/workflow/mutable_state_impl.go [hunk 1]
```diff
@@ -1356,6 +1357,41 @@ func (ms *MutableStateImpl) GetCompletionEvent(
...
+				NextPageToken:          []byte{},
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			for _, event := range resp.HistoryEvents {
+				// this only applies to terminated workflows whose ultimate WFT had been failed
+				if event.EventType == enumspb.EVENT_TYPE_WORKFLOW_TASK_FAILED {
+					return event, nil
```
Comment (yycptt) [inline service/history/workflow/mutable_state_impl.go:1387]: we need to return the workflow completion event, which is the terminated event.

File: service/history/workflow/mutable_state_impl.go [hunk 2]
```diff
@@ -1356,6 +1357,41 @@ func (ms *MutableStateImpl) GetCompletionEvent(
...
 		}
+
+		// Certain terminated workflows have an incorrect completionEventBatchId recorded which
+		// prevents the completion event from being loaded outside of cache.
+		//
+		// If we get back an internal error, attempt to search history (most recent
+		// events first) to find the completion event. Event will be earlier than
+		// the recorded CompletionEventBatchID, and should be a part of the same batch.
+		//
+		// see #inc-791-2024-07-16-high-frontend-cpu-on-s-cd046
```
[the remark this answers is not in the document]
Comment (lina-temporal) [reply service/history/workflow/mutable_state_impl.go:1368]: 👍  sure, will fix

File: service/history/workflow/mutable_state_impl.go (hunk 1 above)
Comment (lina-temporal) [reply service/history/workflow/mutable_state_impl.go:1389]: fixed, thanks!

File: service/history/api/respondworkflowtaskcompleted/api.go [hunk 3]
```diff
@@ -416,8 +419,14 @@ func (handler *WorkflowTaskCompletedHandler) Invoke(
 			// Flush buffer event before terminating the workflow
 			ms.FlushBufferedEvents()
 
-			if err := workflow.TerminateWorkflow(ms, wtFailedCause.Message(), nil,
-				consts.IdentityHistoryService, false); err != nil {
+			_, err := ms.AddWorkflowExecutionTerminatedEvent(
```
Comment (yycptt) [inline service/history/api/respondworkflowtaskcompleted/api.go:422]: Please check if there's any existing functional tests exercising this code path.

**Review by yycptt — APPROVED** [review]

File: service/history/workflow/mutable_state_impl.go [hunk 4]
```diff
@@ -1350,12 +1351,47 @@ func (ms *MutableStateImpl) GetCompletionEvent(
...
+			resp, err := ms.shard.GetExecutionManager().ReadHistoryBranchReverse(ctx, &persistence.ReadHistoryBranchReverseRequest{
+				ShardID:                ms.shard.GetShardID(),
+				BranchToken:            currentBranchToken,
+				MaxEventID:             nextEventID, // looking for an event in the most recent batch
+				PageSize:               1,
+				LastFirstTransactionID: txID,
+				NextPageToken:          []byte{},
+			})
+			if err != nil {
+				return nil, err
```
Comment (yycptt) [inline service/history/workflow/mutable_state_impl.go:1377]: let's catch the not found error from here as well. probably make it a defer function.

**Review by yycptt — APPROVED** [review]

File: service/history/workflow/mutable_state_impl.go (hunk 4 above)
Comment (lina-temporal) [reply service/history/workflow/mutable_state_impl.go:1386]: done

**Review by yycptt — APPROVED** [review]

**Pull request merged**